Direct Agent Execution

CharliePoole opened the pull request:
Fixes #1623

manfred-brands commented on src/NUnitCommon/nunit.agent.core.tests/Runners/DomainManagerStaticTests.cs:
```diff
@@ -16,11 +16,7 @@ public static class DomainManagerStaticTests
-#else
-        const string STANDARD_CONFIG_FILE = "nunit.engine.core.tests.dll.config";
-#endif
+        const string STANDARD_CONFIG_FILE = "nunit.agent.core.tests.exe.config";
```
Compiling this branch, I see the previous behaviour:
```
$ find . -name 'nunit.agent.core.tests*.config'
./net462/nunit.agent.core.tests.exe.config
./net8.0/nunit.agent.core.tests.dll.config
```

manfred-brands reviewed (changes requested):
I don't think this new code is part of the testing as I don't see it in build.cake or package-tests.cake but I'm not clear on how the tester is tested.

CharliePoole commented:
The DirectTestAgent isn't part of the engine but of the tests. It just provides an exe to allow exercising the direct execution facility in nunit.agent.core. There's only one unit test that uses it right now but once the agents are in separate projects, all their tests will be run through it. We have never tested the agents in isolation before but this makes a start.

CharliePoole commented on src/NUnitCommon/nunit.agent.core.tests/AgentDirectRunnerTests.cs:
```diff
@@ -0,0 +1,56 @@
+#if NETFRAMEWORK
+            string agentExe = agentAssembly;
+#else
+            string agentExe = Path.ChangeExtension(agentAssembly, ".exe");
```
It should continue to be supported. Some linux distros support exe but not all. Let's add an issue for this.

CharliePoole commented:
This has been dormant for a while, so I'm making it WIP while I rebase and decide how to move it  forward.

CharliePoole commented:
@manfred-brands

This is again ready for review. I took most of your suggestions and brought the PR up to date with the latest commits. Since that included the new analyzers, I fixed the analyzer problems. I also added a few package tests to actually make use of the feature.

For the moment, I have kept the code duplication for console output. I think some of this code will change once the agents are extensions in their own right. At that point, I'd like to allow it to evolve independently. We'll then have a better idea of exactly what parts can be common and eliminate the duplication.

manfred-brands commented on src/NUnitCommon/nunit.agent.core.tests/Program.cs:
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) Charlie Poole, Rob Prouse and Contributors. MIT License - see LICENSE.txt
+
+using System.Reflection;
```
This using statement is not needed: Removed.

manfred-brands commented on src/NUnitCommon/nunit.agent.core/AgentDirectRunner.cs:
```diff
@@ -0,0 +1,325 @@
+            }
+        }
+
+        private static TAttr? GetAttribute<TAttr>(Assembly assembly)
+            where TAttr : Attribute
+        {
+            return assembly?.GetCustomAttribute<TAttr>();
+        }
```
This method no longer adds any simplification or abstraction: Removed.

manfred-brands commented on src/NUnitCommon/nunit.common/XmlHelper.cs:
```diff
@@ -1,8 +1,10 @@
+using System.Collections.Generic;
 using System.Globalization;
 using System.Xml;
+using System.Xml.Linq;
```
Not needed: Removed.

manfred-brands commented on src/NUnitCommon/nunit.common/XmlHelper.cs:
```diff
@@ -145,5 +149,32 @@ public static DateTime GetAttribute(this XmlNode result, string name, DateTime d
             return date;
         }
+
+        public static IEnumerable<XmlNode> NonNullChildNodes(this XmlNode node)
```
Not used: Removed.

manfred-brands commented on src/NUnitCommon/nunit.common/XmlHelper.cs:
```diff
@@ -145,5 +149,32 @@ public static DateTime GetAttribute(this XmlNode result, string name, DateTime d
+                    yield return childNode;
+        }
+
+        public static void ForEachChildNode(this XmlNode node, Action<XmlNode> action)
```
Could simply call: `node.ChildNodes.ForEachNode(action)`: Updated,

manfred-brands reviewed (commented):
A few small items which I have fixed in a review commit.
The main issue I have is what I said before, why is the reporting different between direct agent execution and the ConsoleRunner.
Both call `runner.Run` and print the results.
Why can the `AgentDirectRunner` not create an `ResultReporter` for the report?

manfred-brands reviewed (approved):
Approved with the caveat of your message that the reporting will be re-visited.

CharliePoole commented:
Thanks for your review. I think it was important to get this odd bit of code in place before I start splitting out all the agents into separate projects. I also agree that something ought to be done about the output before that happens.

The duplication in this PR is duplication I just created _by design_. I copied some code from the console, solely as a way to get us  started. I want people to see this feature and give us feedback with ideas about where it could be used and how it should look in the future.

Once that happens, we might remove the duplication in several ways...
* Dropping the output entirely and relying on the console runner via some option.
* Refactoring so that the report class is somewhere in common and used by both
* Factoring out common pieces while keeping the rest specific to the individual application

If I had decided _not_ to have the duplication in the first place, I would be picking one of those things in advance.

If this sounds crazy to you, that's OK too. It's something I've done in a lot of places and I've had a lot of strange reactions at first, but it seems to pay off more than not. ;-)

CharliePoole commented:
Looks like that alpha.34 label was never removed. I'm a bit tired, so I'll mess directly with the labels on version4 branch in the morning when I'm fresh.

CharliePoole commented:
I got a few hours sleep and fixed the tags and did a push --tags -f on version4. Reran the build and it all looks good now. The problem was that the alpha.34 and alpha.35 tags were on detached commits. 

Latest on myget is now 4.0.0-alpha.35 and I think we're now good for future runs.
